[Aw Ming Yi] iP

MingYiAw opened the pull request:
## Dude

> "Catchphrase!" –  Dude, Free Guy 2021

Dude frees your mind of having to remember things you need to do. It's,

- dark mode enabled
- text-based
- easy to learn
- ~~SUPER~~ EXTREMELY FAST to use

All you need to do is,
1. download it from [here](https://github.com/MingYiAw/ip/releases/tag/v0.1)
2. run `java -jar dude.jar` from the directory
3. add your tasks
4. let it manage your tasks for you 💪

 And it is **FREE**!!!

Features:

- [ ]  Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

MugenNoKyosu committed:
no message

MugenNoKyosu committed:
Revert "no message"

This reverts commit f548fbfb019cec680575d4b14f2a7d000f5ca73e.

MugenNoKyosu committed:
Rename, Greet, Exit

MugenNoKyosu committed:
add echo

MugenNoKyosu committed:
no message

MugenNoKyosu committed:
add Task class

MugenNoKyosu committed:
add Task class

MugenNoKyosu committed:
add MarkAsDone

MugenNoKyosu committed:
no message

MugenNoKyosu committed:
add ToDo, Deadline, Event classes

MugenNoKyosu committed:
add method to add different task to list

MugenNoKyosu committed:
set up test

MugenNoKyosu committed:
debug Deadline and Event output

MugenNoKyosu committed:
add exceptions

MugenNoKyosu committed:
add throw exceptions in dude

MugenNoKyosu committed:
debug Deadine and Event output

MugenNoKyosu committed:
add method to delete task

MingYiAw committed:
Adjust coding style to meet course coding standards

MingYiAw committed:
Sort classes into package

MingYiAw committed:
Add DudeCorruptedDataException.java

The program may crashes when encountering corrupt data while reading the text file.

Let's add DudeCorruptedDataException class to handle corrupt data during file reads.

Using exception in this case ensure the program can continue operating instead of failing.

MingYiAw committed:
feat: add data saving system

The program lacks the ability to save task, resulting in data loss when the exiting program.

Data need to be persisted and able to restore when user restarts the program.

Let's add a saving system to store user data in a text file.

Text file is used to store the data for easy access.

MingYiAw committed:
Merge branch 'branch-Level-7'

MingYiAw committed:
fix: remove redundant try-catch in exit() method

There is a try-catch block in exit() method which handle exception throw
by saveData() method.

The saveData() method is being refined and no longer throw any
exception, DudeDataFileException class is also deleted, these cause
the try-catch block in exit() method become redundant and cause
compilation error.

Let's remove the try-catch block for saveData() method in exit() method.

MingYiAw committed:
feat: add dates and times check

Users are able to input anything as date and time while creating a
new task.

Program able to recognize valid date and time is more user-friendly
as it helps user to create tasks with valid and correct details.

Let's,
* define a date and time format for user input
* update addDeadline() and addEvent() methods to check for valid
date and time
* add a new exception named DudeDateTimeFormatException to
deal with invalid date and time input

Using a fixed format for input enable the program to recognise valid
input easily and the new exception is used to inform user on the
format error and show user the correct format.

MingYiAw committed:
refactor: change DudeCorruptedDataException and DudeNullDateTimeException message

Exceptions' message change to make it more clear and precise.

MingYiAw committed:
feat: add validation to ensure 'from' date is before 'to' date

Users are able to input any date and time for 'from' and 'to' when creating an event as long as the format is correct.

The program allows users to create an event with invalid time range.

Let's implement a check to validate that the 'from' date is earlier than the 'to' date.

MingYiAw committed:
Merge branch 'branch-Level-8'

MingYiAw committed:
feat: Add CommandType.java

The application does not have an enumeration to represent the different
types of commands that can be issued.

Without an enumeration, command handling is less organized and
make code more difficult to understand.

Let's add an enum named CommandType to represent the different
types of commands.

Using an enum ensures that all command types are centrally defined,
making the code more maintainable and easier to understand.

MingYiAw committed:
Merge branch 'branch-A-MoreOOP'

MingYiAw committed:
refactor: Extract functionalities from Dude.java into seperate classes

Several functionalities are packed in Dude class such as user
interaction, file handling, command parsing and task management.

This tight coupling makes the code difficult to maintain and
understand, which can lead to potential issues in future development.

Let's extract related functionalities into separate classes such
as Ui, Storage, Parser, and TaskList.

This enhances code readability, maintainability and also adhering to
the Single Responsibility Principle.

MingYiAw committed:
docs: Add Javadoc for classes and methods

There is no Javadoc for many classes.

Without proper documentation, the code is difficult to understand and maintain,
especially for new developers or contributors.

Let's add Javadoc comments to all classes and methods.

Clear Javadoc improves code readability and aids future development.

MingYiAw committed:
Merge branch 'branch-A-MoreOOP'

MingYiAw committed:
refactor: Put Task related classes into package dude.task

MingYiAw committed:
Merge branch 'add-gradle-support'

MingYiAw committed:
fix: Set up Gradle to run with correct class name

MingYiAw committed:
refactor: Update DudeDateTimeFormatException message

MingYiAw committed:
test: Add JUnit test to few classes

The project currently lacks automated testing.

Adding automated tests ensures the correctness of the code.

Let's add JUnit tests for few classes such as Parser, TaskList and Deadline.

MingYiAw committed:
fix: Set up Gradle to create JAR file

MingYiAw committed:
ci: Set up Gradle to use checkstyle

MingYiAw committed:
refactor: Adjust coding style to meet course coding standards

MingYiAw committed:
Merge branch 'branch-A-CodingStandard'

MingYiAw committed:
feat: Add find command

Current system lacks a way to search for tasks by a specific keyword.

Search function can enhance user experience and make task management more efficient.

Let's add a find command to search and filter tasks based on a given keyword.

MingYiAw committed:
test: Add test cases to test findAllTask method in TaskList

MingYiAw committed:
fix: Update find method to handle empty task descriptions

Find command will react as no match found when no searching
keyword is provided.

Having a different message for no description and no search result
for find command enchance user experience because users are
able to see the problem clearly.

Let's update find method to throw DudeNullDescriptionException
when no description is provided.

MingYiAw committed:
Merge branch 'branch-Level-9'

kvernnn commented on src/main/java/dude/task/Deadline.java:
```diff
@@ -0,0 +1,43 @@
+public class Deadline extends Task {
+    protected LocalDateTime by;
+
+    /**
+     * Constructs a Deadline with the specified description and deadline.
+     * Date and time must be in specific format: "yyyy-MM-dd HH:mm"
+     *
+     * @param description The description of the task.
+     * @param by The date and time represent the deadline.
+     */
```
I really like how you clearly explain what the constructor does in each of your classes.

kvernnn commented on src/main/java/dude/task/TaskList.java:
```diff
@@ -0,0 +1,112 @@
+package dude.task;
+
+import java.util.ArrayList;
+
```
Should there be an empty line here?

MingYiAw replied:
Thanks for the reviews. 
I think it's needed here according to the coding standard, since `java.util.ArrayList` and `dude.exception.DudeDuplicatedTaskException` belongs to different import groups.

kvernnn commented on src/main/java/dude/exception/DudeTaskNotFoundException.java:
```diff
@@ -0,0 +1,14 @@
+ * Thrown to indicate that no task with matching prompt is found in TaskList.
+ */
+public class DudeTaskNotFoundException extends DudeException {
+
```
Should there be an empty line here? Spotted this in other areas as well.

MingYiAw replied:
This isn't needed, but out of habit, I tend to leave a blank line before any method. It doesn't violate any rules when running `checkstyleMain`. I hadn't noticed this before., not sure whether this should be avoided or not.

kvernnn commented on src/main/java/dude/Ui.java:
```diff
@@ -0,0 +1,156 @@
+    /**
+     * Displays the greeting message.
+     */
+    public void showGreet() {
+        System.out.println(LINE);
+        System.out.println("Hello! I'm " + BOT_NAME);
+        System.out.println("What can I do for you?");
+        System.out.println(LINE);
+    }
```
I really like how your methods follow the coding standard. i.e method name is a verb

kvernnn commented on src/main/java/dude/task/Task.java:
```diff
@@ -0,0 +1,63 @@
+ */
+public abstract class Task {
+    protected String description;
+    protected boolean isDone;
```
I think it is really good that your booleans follow the coding standard. i.e starts with "is"

kvernnn reviewed (commented):
LGTM, might require some cosmetic changes, I might be wrong though.

MingYiAw committed:
feat: Add GUI with JavaFX

The application currently operates through Text UI, which limits its
user experience.

GUI can enhance user experience by making the application more
visually appealing and user-friendly.

Let's implement a GUI for Dude.java.

Using JavaFX to create a GUI allows for seamless integration with
the existing codebase.

MingYiAw committed:
refactor: Put GUI related classes into package dude.gui

MingYiAw committed:
fix: Fix application unable to close after "bye" command

The application remains open after using the 'bye' command due to the
lack of proper shutdown logic in the GUI.

Users expect the application to close when they use the "bye" command,
the lack of this functionality could causes confusion and a poor user
experience.

Let's,
* add a method in Dude to check if it is running
* add method in MainWindow to close the application after a short delay if
Dude is no longer running

The short delay ensures that application display the exit message before
the it exits.

MingYiAw committed:
test: Correcting testcases for ParserTest and TaskListTest

MingYiAw committed:
feat: Change Send Button hover and pressed color

MingYiAw committed:
Merge branch 'branch-Level-10'

SlothyCat commented on src/main/java/dude/CommandType.java:
```diff
@@ -0,0 +1,18 @@
+ * Representing the different types of commands that can be issued.
+ * Each command corresponds to a specific action within the application.
+ */
+public enum CommandType {
```
Great Attempt to make use of Enums 😃

SlothyCat commented on src/main/java/dude/Dude.java:
```diff
@@ -0,0 +1,310 @@
+    /**
+     * Checks if the Dude application is currently running.
+     *
+     * @return True if the Dude application is running, false otherwise.
```
Very clear comment indicating the reason for each boolean values 😺

SlothyCat commented on src/main/java/dude/task/ToDo.java:
```diff
@@ -0,0 +1,36 @@
+package dude.task;
+
+/**
```
Very thorough and detailed comments even for a small and simple class. YOURE TOO GOOD

SlothyCat reviewed (commented):
YOU ARE TOO GOOD 😆

MingYiAw committed:
refactor: Add assertions to ensure internal logic consistency

No assertion used in any class.

Assertions are important to catch bugs during development and ensure
application runs as expected.

Let's,
* add assertion to check that file path is not empty
* add assertion to check status of Dude after running each method

MingYiAw committed:
refactor: Restructure some methods in Dude.java

Some redundant else statement is used in several methods.

Removing extra else statement make the code more neat and
increase the code readability.

Let's remove redundant else statement in Dude.java methods.

Instead of using a full if-else statement, guard clauses are
implemented to make the happy path clear and increase the
code readability.

MingYiAw committed:
refactor: Add stream to filter tasks

MingYiAw committed:
Merge pull request #1 from MingYiAw/branch-A-Assertions

refactor: Add assertions to ensure internal logic consistency

MingYiAw committed:
Merge branch 'master' of https://github.com/MingYiAw/ip

* 'master' of https://github.com/MingYiAw/ip:
  refactor: Add assertions to ensure internal logic consistency

MingYiAw committed:
Merge branch 'master' into branch-A-CodeQuality

MingYiAw committed:
Merge branch 'master' into branch-A-Streams

MingYiAw committed:
Merge pull request #2 from MingYiAw/branch-A-CodeQuality

refactor: Restructure some methods in Dude.java

MingYiAw committed:
Merge branch 'master' into branch-A-Streams

MingYiAw committed:
Merge pull request #3 from MingYiAw/branch-A-Streams

refactor: Add stream to filter tasks

MingYiAw committed:
ci: Set up Continuous Integration using GitHub action

The project does not have automated CI set up.

Automated testing through CI ensures code quality and verifies changes
before merging.

Let's use GitHub Actions to set up CI to run on each commit and pull
request.

MingYiAw committed:
fix: Fix some merging bug in Dude.java

MingYiAw committed:
style: Remove unuse import in TaskList.java

MingYiAw committed:
feat: Add user define shortcut for commands

Users need to type full commands for Dude to understand.

Typing out full commands is time consuming and affects user efficiency,
this violates the feature of the application for being fast to use.

Let's,
* Add command to define custom shortcuts for each commands
* Add command to remove previously defined shortcuts

Custom shortcuts reduce typing effort and streamline the user's workflow,
offering a better user experience.

MingYiAw committed:
docs: Add and update Javadoc for user define shortcut update

MingYiAw committed:
docs: Update defineShortcut() Javadoc to follow standard style

MingYiAw committed:
Merge branch 'branch-C-FriendlierSyntax'

MingYiAw committed:
refactor: Add assertions for define() and undefine()

MingYiAw committed:
style: Update ParserTest to follow coding standard

MingYiAw committed:
feat: Improve GUI

Current GUI has a standard font and text size, and lacks visual
feedback for errors.

Improving these aspects enhance user experience and provide
clearer visual cues for errors.

Let's,
* Change the font style
* Increase the font size
* Crop profile picture into circle
* Make profile picture glows in red when displaying error
message
* Display error message in red color and bolded

Improving error display on both profile picture and text
enhance error visibility.

MingYiAw committed:
Merge branch 'branch-A-BetterGui'

MingYiAw committed:
feat: Rewrite few reply of Dude

Dude's messages are bland and boring.

Changing Dude's reply into something more interesting can
enchance user experience.

Let's modify Dude's phrases according to a personality.

Adding personality to Dude improve the quality of the application
and enchance user experience.

MingYiAw committed:
feat: Rewrite Dude's error messages

Dude's error messages doesn't match the personality of Dude which is
decided earlier.

Rephrasing Dude's error messages to match its other messages
shows the completeness of the application.

Let's modify Dude's error messages according to its personality.

MingYiAw committed:
Merge branch 'branch-A-Personality'

MingYiAw committed:
style: Refine methods to follow coding standard

MingYiAw committed:
fix: Add DudeInvalidNameException.java

Users can add task with description that contains characters
that leads to data corruption.

The occurence of data corruption diminish user experience and
reduce the usefulness of the application.

Let's add DudeInvalidNameException to deal with invalid task
name.

Using exception to deal with this issue can prevents data
corruption and inform the user about it.

MingYiAw committed:
Merge branch 'branch-A-MoreErrorHandling'

MingYiAw committed:
Merge branch 'master' into branch-A-MoreErrorHandling

MingYiAw committed:
test: Rewrite and add testcases for multiple class

There are test cases available for only three classes, and most are
outdated due to recent updates.

The existing tests do not reflect the latest code changes, which may
lead to incomplete or inaccurate testing coverage.

Let's,
* Add new test cases for ToDo.java and Event.java
* Rewrite outdated test cases to align with recent updates

Ensuring the tests are up-to-date and comprehensive improves
overall code quality and reliability.

MingYiAw committed:
Merge branch 'branch-A-MoreErrorHandling'

MingYiAw committed:
docs: Rewrite and add Javadoc for multiple classes

Few classes are missing proper Javadoc or had outdated Javadoc.

Lack of proper Javadoc makes the code harder to understand and
to maintain.

Let's,
* Rewrite outdated Javadoc for few methods in Dude.java
* Add Javadoc for new classes and methods

This was done to improve code readability and maintainability,
making the code more accessible to other developers.

MingYiAw committed:
docs: Add user guide

Current application lacks proper documentation for users.

Lack of user guide will cause confusion and greatly diminish user
experience.

Let's add a user guide as README.md

This guide is structured to help users quickly understand how to
interact with the application, improving overall user experience.
